### ISSUE TITLE
Factual error in `PartialEq` documentation

### ISSUE THREAD

**Issue by QuineDot** [rust-lang/rust#91843]
[The documentation for `PartialEq`](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html) states:
> `x.eq(y)` can also be written `x == y`, and `x.ne(y)` can be written `x != y`.

This is not true, as #44762 illustrates, as well as this [URLO](https://users.rust-lang.org/t/why-can-we-not-compare-values-references/68709)-inspired [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9623954d3bfbe3c1bf98078b21802fc9):
```rust
    let val = 1i32;
    let val_ref = &val;
    let is_equal = val.eq(val_ref); // Compiles
    let is_equal = val == val_ref; // Does not compile    
```

The other operator's documentation don't have this error.  [`Shl`](https://doc.rust-lang.org/std/ops/trait.Shl.html) (and `Shr`) also explicitly note some differences in inference between operators and the method call.  

@rustbot label +A-docs

**Comment by camelid** [comment]
I assume this is because auto(de)ref kicks in for the method version but not the operator version.

**Comment by QuineDot** [comment]
> I assume this is because auto(de)ref kicks in for the method version but not the operator version.

Yes (and experimenting with changing that is what #44762 is about).

**Comment by JanBeh** [comment]
The [reference on comparison operators](https://github.com/rust-lang/reference/blob/2cb0ed9ba56360949f492f9866afe8c293f9f9da/src/expressions/operator-expr.md#comparison-operators) clarifies:

> Unlike the arithmetic and logical operators above, these operators [the comparison operators] implicitly take shared borrows of their operands, evaluating them in [place expression context](https://github.com/rust-lang/reference/blob/2cb0ed9ba56360949f492f9866afe8c293f9f9da/src/expressions.md#place-expressions-and-value-expressions):
> 
> ```rust
> # let a = 1;
> # let b = 1;
> a == b;
> // is equivalent to
> ::std::cmp::PartialEq::eq(&a, &b);
> ```
> 
> This means that the operands don't have to be moved out of.

**Comment by QuineDot** [comment]
[But that's also counterfactual!](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5ca19bd0f237ed9c22585a78025b1cd9)
```rust
// Commented code does not compile
fn main() {
    let v = 1;
    let x: *const u32 = &v;
    let y: &u32 = &1;
    let _ = dbg!(x == y);
    // let _ = dbg!(PartialEq::eq(x, y));
    // let _ = dbg!(PartialEq::eq(x, &y));
    // let _ = dbg!(PartialEq::eq(&x, y));
    // let _ = dbg!(PartialEq::eq(&x, &y));
    let _ = dbg!(PartialEq::eq(&x, &(y as _)));

    // But that can't be a desugaring...
    // let _ = dbg!(v == y);
    let _ = dbg!(&v == y);
    // let _ = dbg!(PartialEq::eq(&v, &(y as _)));
}
```

Operator resolution is its own thing with many niche differences and, in my opinion, all phrasing like "equivalent to" / "can be written as" / "is the same as" / "desugars" should be replaced with something more like "is notionally" / "is like" / etc, perhaps with a list of exceptions you're likely to hit.

Another example from the reference is [compound assignment (`+=`):](https://doc.rust-lang.org/reference/expressions/operator-expr.html#compound-assignment-expressions)
* They mention a place expression requirement -- great!
* They mention order evaluation differs for primitives(!) -- great!
* But then they blunder and say
  > the following expression statements in `example` are equivalent:

Even modulo the exceptions they did cover, [the remaining cases are not equivalent because `+=` can participate in two-phase borrowing.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1a686caae3dc5a5f83f8b8997e632ed0)
```rust
// commented code does not compile
use core::num::Wrapping;
fn main() {
    let n = &mut [Wrapping(0)][..];
    n[0] += n[0];
    // std::ops::AddAssign::add_assign(&mut n[0], n[0]);
}
```
([See also.](https://github.com/rust-lang/rust/issues/72199))

**Comment by hkBst** [comment]
@QuineDot it seems to me that you have accumulated enough knowledge around this issue to create a good PR for this. The documentation lives here: https://github.com/rust-lang/rust/blob/master/library/core/src/cmp.rs.
